### ISSUE TITLE
[Question] Adequate path of bank collaborator's computer to place and install Cmder. Or ConEmu?

### ISSUE THREAD

**Issue by gusbemacbe** [cmderdev/cmder#2640]
### Purpose of the issue
- [ ] Bug report (encountered problems/errors)
- [ ] Feature request (request for new functionality)
- [X] Question

### Version Information
- Cmder 1.3.18
- Windows 10 20H2

### Description of the issue
I am totally aware that Cmder and ConEmu are totally similar, while Cmder does not have a standalone installer and ConEmu does. But I am not sure if ConEmu comes with all Cmder featureas Clink, Git/MinGW and clink-completions. 

I read the documentation that Cmder should not be placed on the path `C:\Program Files\`, but ConEmu can be installed on this path. 

Then I open a call to demand the addition of Cmder to the bank's its own software central so any bank developers can download and install. The Cmder addition demand was validated and approved. Since you said it can not be installed on the path `C:\Program Files\`, I suggested the support analyst to set the path to `%APPDATA%`, which does not require administrator. It was shipped to the software central. I tested and it gave an installation error. 

The support analyst explained it worked in the computer, but didn't work in the software central because of `%APPDATA%`, so the software central's standard path always is `C:\Program Files\`. To work on the software central, it needs to be on `C:\Program Files\`.

Any solution?

I saw the issue #2552 and @edrohler's plan to make a standalone installer of Cmder, but he abandoned it. 

If you answer ConEmu comes with all above cited features, we'll not consider Cmder and we'll replace with ConEmu, but if ConEmu does not, then we would like to use Cmder on bank computers, but it needs to be placed on that path so the software central and the installation would work without fails.

**Comment by gusbemacbe** [comment]
Cc: @Maximus5

**Comment by chrisant996** [comment]
@gusbemacbe Cmder stores writable data files under its program directory.  But Administrator access is necessary to be able to write to data files under `C:\Program Files\`.  That's why Cmder doesn't belong in `C:\Program Files\` -- it would require running as Administrator in order to use Cmder, which is not a very secure way to operate (especially on a bank computer).

> If you answer ConEmu comes with all above cited features, we'll not consider Cmder and we'll replace with ConEmu, but if ConEmu does not, then we would like to use Cmder on bank computers, but it needs to be placed on that path so the software central and the installation would work without fails.

ConEmu only has the ConEmu features.  ConEmu does not include the Cmder integration features or Clink or clink-completions or git or custom prompts for cmd/powershell/etc.  That is why Cmder exists.

Ed's installer apparently worked, but there are two problems:
1. Ed made the installer with the intent of adding Cmder to winget.  The reason it was discontinued is because winget wants software from the original source, and the installer was not from the software's original source.
2. The installer still cannot install Cmder to `C:\Program Files\` because of where Cmder stores its data files.  This is an issue that would require changes in Cmder itself to support separating writable data files from program files.

**Comment by gusbemacbe** [comment]
@chrisant996 

> Cmder stores writable data files under its program directory. But Administrator access is necessary to be able to write to data files under `C:\Program Files`\. That's why Cmder doesn't belong in `C:\Program Files\` -- it would require running as Administrator in order to use Cmder, which is not a very secure way to operate (especially on a bank computer).

Too bad.

> ConEmu only has the ConEmu features. ConEmu does not include the Cmder integration features or Clink or clink-completions or git or custom prompts for cmd/powershell/etc. That is why Cmder exists.

So somebody in the bank's IT area has to do manually to include Cmder features to ConEmu, according to the documentations – https://conemu.github.io/en/GitForWindows.html and https://conemu.github.io/en/TabCompletion.html. Would they be able to include these features to ConEmu installed in `C:\Program Files\`? And somewhere?

**Comment by chrisant996** [comment]
Just to be clear -- you are talking about installing multiple programs, not "features".

Any program you want to install that already has an installer that supports being installed under `C:\Program Files\` should be fine because the program already meets the requirements.

The [Clink](https://github.com/chrisant996/clink) program can be installed to `C:\Program Files\`.
And git can be installed to `C:\Program Files\`.

The Cmder package includes a few other things, too, though.  Such as [clink-completions](https://github.com/vladimir-kotikov/clink-completions) and a pre-customized prompt for cmd/clink and powershell, etc.

Clink can be told to load scripts from wherever you like, so if people are able to copy .lua script files to some personal files location on a bank computer then they could copy .lua scripts on the computers themselves.

Or someone within the bank could create an installer for whatever scripts you want to include by default.  E.g. [clink-completions](https://github.com/vladimir-kotikov/clink-completions) and [clink-flex-prompt](https://github.com/chrisant996/clink-flex-prompt) might be two you'd be interested to consider.

**Comment by daxgames** [comment]
`cmder.exe -c [folderpath]` can be run to tell cmder to store writeable config in `[folderpath]`.  It just require a custom shortcut.

`-c [folderpath]` should normally be in a user home folder like ` -c "%userprofile%\cmder_config`.

**Comment by daxgames** [comment]
`-c [folderpath]` also enabled a kind og shared config mode where a system administrator can store config in `%CMDER_ROOT%\config` that is processed before `[folderpath]\config`.

This allows the sysadmin to setup defaults and allows the user to have his/her own config that begins with the sysadmins dedaults.

You can search the issues for how other users are using this feature.

**Comment by chrisant996** [comment]
@gusbemacbe so the `/c` info above means that if someone builds an installer for Cmder, it will be possible to install to `c:\program files\` as long as the installer or the user makes a shortcut that launches Cmder using the `/c` flag to point it at a configuration folder that is in a normal directory (i.e. that does not require Administrator access like the Program Files directories do).

**Comment by chrisant996** [comment]
> Keep on mind this voids one of the core principles of cmder as a portable app since user config is no longer stored with the binaries.

There is a difference between "supporting portable install" versus "prohibiting non-portable install".

While a portable install can be great when the user wants a portable install, not all users require that.  And users whose IT departments prohibit it are then blocked from using Cmder.

For example, Clink supports a fully portable install (there is section in the docs specifically discussing it).  Clink also supports a non-portable install.  Which empowers the user to choose which approach best meets the user's needs.

**Comment by daxgames** [comment]
@chrisant996 Exactly.  This is why we added this capability.

**Comment by gusbemacbe** [comment]
@chrisant996 

> The [Clink](https://github.com/chrisant996/clink) program can be installed to `C:\Program Files\`. 
> And git can be installed to `C:\Program Files\`.

It is a good sign. I'll talk with the support analyst. 

> But you can probably add your own customizations instead. For example, Clink can be told to load scripts from wherever you like, so if people are able to copy .lua script files to some personal files location on a bank computer then they could copy .lua scripts on the computers themselves.

I'll explain how the bank restrictions work. 

Cmder and .lua scripts have to be shared among squads or colleagues, that have them, under their manager's certification and approval, and shared and stored on a bank's Confluence site since EPM, PMS and cybersecurity teams approve, certificate, unblock (if the app is not stored on the Confluence page, but downloaded from the Cmder site) and releases.

But not all squads and users will have it. It is why the bank's own software central exists so any user can download and install without asking the colleagues or another squad and why the bank's Confluence site exists so the users can catch, download the stored files or from the unblocked sites and install, outside the bank's software central, with enabled EPM, without asking their colleagues and other squads, and without asking the permission of their manager, coordinator, superintendent, and PMS or cybersecurity people. 

> Or someone within the bank could create an installer for whatever scripts you want to include by default. E.g. [clink-completions](https://github.com/vladimir-kotikov/clink-completions) and [clink-flex-prompt](https://github.com/chrisant996/clink-flex-prompt) might be two you'd be interested to consider.

Another good sign. I'll open new calls with ConEmu and these programs for being added to the bank's software central. I'll inform the support analyst. 

@daxgames and @chrisant996 

> `/c [folderpath]` also enables a kind of shared config mode where a system administrator can store config in `%CMDER_ROOT%\config` that is read only to the user and processed before the user specific `[folderpath]\config`.
> 
> This allows the sysadmin to setup defaults and allows the user to have his/her own config that begins with the sysadmins defaults.
> 
> You can search the issues for how other users are using this feature.

> @gusbemacbe so the `/c` info above means that if someone builds an installer for Cmder, it will be possible to install to `c:\program files\` as long as the installer or the user makes a shortcut that launches Cmder using the `/c` flag to point it at a configuration folder that is in a normal directory (i.e. that does not require Administrator access like the Program Files directories do).

The support analyst did it. Cmder on `C:\Program Files\` always asks the user to answer the administrator password, what is prohibited by the bank. 

> While a portable install can be great when the user wants a portable install, not all users require that. And users whose IT departments prohibit it are then blocked from using Cmder.

The bank allows the users to share each other and to use Cmder since it be certified and approved by their superintendent, coordinator and manager. It can be included on the list of apps, libraries and packages approved, certificated, unblocked and released by the EPM team on the Confluence page to store or not, catch and download the stored file or from the unblocked Cmder site outside the bank's software central. 

With enabled EPM, unfortunately Cmder and ConEmu sites are still not released to download and install the programs, it will take a lot of time to approve, certificate, unblock and release, while the sites of Anaconda, Docker, Git Bash, JetBrains, Nodejs, Python, Spotify, Sublime Text, Terraform and VSCode were approved, unblocked and released to download and install outside the bank's software central. 

I'll talk with and try to persuade the EPM team to analyse and consider it. 

All above these cited programs, with the exception of Spotify and Terraform, that the others demanded, were approved and added to the bank's software central. PowerToys, Windows Terminal and Tabby that I demanded, were approved early and added to the bank's software central. Only Cmder failed. I'll open a call with ConEmu and other programs cited by you. 

- As ConEmu has an installer, it should be approved, added to the bank's software central, and released with EPM to download and install outside the bank's software central. For the bank software central, it will take a bit of time to approve, validate, certificate and add. For the EPM, it will take a lot of time to approve and release. 
- As Cmder is a portable program, it should be approved and released with EPM to store on the Confluence page, download the stored file or from the site and install outside the bank's software central. For the EPM, it will take a lot of time to approve and release.

**Comment by chrisant996** [comment]
> @chrisant996 Exactly.  This is why we added this capability.

But an installer for Cmder is not currently available, right?

Without an installer program, there is no way for a site (such as this bank) to use Cmder with a managed installation approval system.

My understanding was that the reason there is no installer is because it violates the principle of a portable install.

My intent is to suggest that the existence of an installer does not violate the portability principle, and it  allows a user to choose whether to install as portable or not.

**Comment by daxgames** [comment]
If the bank wants an installer they can write one.  Its a simple script that delivers the extracted blessed content of a zip archive they create that is downloaded from their confluence or wherever they decide to store it with their customizations to `c:\program files\cmder` that also places a pre-configured shortcut on the start menu that runs `c:\program files\cmder\cmder.exe /c "%userprofile%\cmder_config"

**Comment by chrisant996** [comment]
> > @gusbemacbe so the `/c` info above means that if someone builds an installer for Cmder, it will be possible to install to `c:\program files\` as long as the installer or the user makes a shortcut that launches Cmder using the `/c` flag to point it at a configuration folder that is in a normal directory (i.e. that does not require Administrator access like the Program Files directories do).
> 
> The support analyst did it. Cmder on `C:\Program Files\` always asks the user to answer the administrator password, what is prohibited by the bank.

I think that "The support analysis did it" means they created a custom installer, and ran the installer, and the installer required Administrator access in order to _install_ to `C:\Program Files\`.

Or do you mean installation worked fine but _running_ Cmder required Administrator access?  (I don't think that's what you mean; that isn't how it should work, if the installer installed it correctly.)

Anyway, on Windows **ALL** programs require Administrator access to install under `C:\Program Files\`.  It is a security restriction that is part of the Windows operating system itself.

@gusbemacbe My understanding is that you've stated the bank doesn't allow users to have Administrator access, but also that the bank requires all programs be installed under `C:\Program Files\` without administrator access.  If that's accurate, then it means programs can never be installed.

Unless the bank has special software that rans as Administrator and installs only certain approved programs.  And if that's the case, then the support analysis would indeed encounter an Admin prompt, because the support analysis was installing it manually, instead of using the automated approved installation system.

- The installer for ConEmu requires Administrator access in order to install under `C:\Program Files\`.
- The installer for Clink currently always requires Administrator access.  I could invest some time and effort to make Administrator access optional to launch the installer, but Administrator access will still be required in order to install under `C:\Program Files\`.

> ... were approved, unblocked and released to download and install outside the bank's software central.

What does it mean to "install outside the bank's software central"?
Does that mean installing outside of `C:\Program Files\`?

If that's an option, then I don't understand what problem exists.  Installing outside of `C:\Program Files\` is the recommended approach.

> * As ConEmu has an installer, it should be approved, added to the bank's software central, and released with EPM to download and install outside the bank's software central. For the bank software central, it will take a bit of time to approve, validate, certificate and add. For the EPM, it will take a lot of time to approve and release.
> * As Cmder is a portable program, it should be approved and released with EPM to store on the Confluence page, download the stored file or from the site and install outside the bank's software central. For the EPM, it will take a lot of time to approve and release.

I'm very confused how these programs are getting installed to `C:\Program Files\`.  Does the bank have special software that is able to automatically install certain pre-approved software using Administrator access, and therefore install under `C:\Program Files\`?

**Comment by daxgames** [comment]
I just tested the suggestion from my last comment.  I am an admin on my PC.

I am assuming that an administrator or some system agent with permissions to install to `c:\program files\cmder` will install Cmder.  Creating a shortcut with the below config:

![image](https://user-images.githubusercontent.com/7318053/147031053-f7af868b-1756-4503-ac8c-a19b14b7a108.png)

Allowed me to launch Cmder.  I received 2 UAC prompts and no password prompts and no additional files were created/written to under `c:\program files\cmder`.  All user config files were written to `%userprofile%\cmder_config`

![image](https://user-images.githubusercontent.com/7318053/147031291-89d786c5-5438-43e6-ad60-4ea471c91a4e.png)

Cmder started as shown below:

![image](https://user-images.githubusercontent.com/7318053/147031425-3b6bdaca-7dc3-4c85-b834-1d4e2802a0fe.png)

I am not sure what the `Access is denied.` message is nor the `The system cannot find the path specified.` but I will look into it.

**Comment by daxgames** [comment]
The errors on startup above are benign.  They are due to a missing folder `c:\program files\cmder\config\profile.d`.  Creating the folder clears the error.

I will add this folder to future Cmder releases with a `README.md` file in it explaining its use.

The UAC prompts appear to be coming from ConEmu and the Cmder team does not have control of this.

**Comment by gusbemacbe** [comment]
> If the bank wants an installer they can write one. Its a simple script that delivers the extracted blessed content of a zip archive they create that is downloaded from their confluence or wherever they decide to store it with their customizations to `c:\program files\cmder` that also places a pre-configured shortcut on the start menu that runs `c:\program files\cmder\cmder.exe /c "%userprofile%\cmder_config"

I will ask the support analyst to test. 



> I think that "The support analysis did it" means they created a custom installer, and ran the installer, and the installer required Administrator access in order to _install_ to `C:\Program Files\`.
> 
> Or do you mean installation worked fine but _running_ Cmder required Administrator access? (I don't think that's what you mean; that isn't how it should work, if the installer installed it correctly.)

I mean the second paragraph. 

> @gusbemacbe My understanding is that you've stated the bank doesn't allow users to have Administrator access, but also that the bank requires all programs be installed under `C:\Program Files\` without administrator access. If that's accurate, then it means programs can never be installed.
> 
> Unless the bank has special software that rans as Administrator and installs only certain approved programs. And if that's the case, then the support analysis would indeed encounter an Admin prompt, because the support analysis was installing it manually, instead of using the automated approved installation system.

In the first paragraph, it is why the bank software central exists to install the approved software with its own administrator privilege. 

> ... because the support analysis was installing it manually, instead of using the automated approved installation system.

Exactly, it was another test he did. 

He did two tests in another bank machine of his with restrictions:

- if Cmder is set on `C:\Users\user\AppData\Local\Programs`, the bank software central would fail to install it. 
- if Cmder is set on `C:\Program Files`, the bank would successfully install it, but when you open Cmder, you'll receive administrator password prompt dialogue/window. 

> What does it mean to "install outside the bank's software central"?
> Does that mean installing outside of `C:\Program Files\`?

Yes, it is this. The bank computers users can download any approved software stored from Confluence site or from any approved, unblocked and released official sites, install on `C:\Users\user\AppData\Local\Programs` outside the bank software central. 

I'll explain how it works:

The EPM team analysed the Microsoft's VSCode site, certificated Microsoft's VSCode site proxy (I mean URL, subdomain addresses) and the files in GZ, ZIP, DMG and EXE formats, approved, unblocked these formats and released the addresses. Then the user needs to sign the responsibility term at the employment contract, to install an EPM program from the bank software central, enables EPM, reboots the computer. With enabled EPM, the users can go to VSCode's official site, downloads an EXE and installs it on AppData without administrator privilege. 

> If that's an option, then I don't understand what problem exists. Installing outside of C:\Program Files\ is the recommended approach.

Yes, the support analyst is not part of the EPM team, he is part of PMS and cybersecurity team. Then I have to talk directly with the EPM team. 

> I'm very confused how these programs are getting installed to `C:\Program Files\`. Does the bank have special software that is able to automatically install certain pre-approved software using Administrator access, and therefore install under `C:\Program Files\`?

Yes, I have explained too much above. Think of Microsoft Store, Google Store and of Apple Store. They do not prompt the administrator password dialogue to install, because they have already the administrator access. 

@daxgames 

As for the last comments, great observations. I'll inform the support analyst.

**Comment by daxgames** [comment]
Recomendations for multi-user redistributable package creation:

1.  Download the Cmder distribution package from Github.
2.  Expand the package to a user writeable folder. Ex: "%userprofile%\cmder"
3.  Launch cmder.exe /c "%userprofile%\cmder_config".
4.  Customize the shared install in:
      - "%userprofile%\cmder\config"
        - Modify `.\user_profile.[cmd|ps1|sh] with session init settings.
      - Add separate `*.[cmd|ps1|sh]` files to `%userprofile%\cmder\config\profile.d` - Each will be incorporated into each shell at shell launch.  If this folder does not exist create it.
      - "%userprofile%\cmder\bin"
        - Add other portable exe files here - This is added to the path.
      - "%userprofile%\cmder\opt"
        Add other software here. Things here are not added to the path but you can in "%userprofile%\cmder\config\user_profile.[cmd|ps1|sh]"
5. Open the Conemu `Setup Tasks` dialog select `Update` from the tree on the left and uncheck startup on the right.
6. Create a Windows shortcut in "%userprofile%\Cmder\Cmder.lnk" configured as follows:

![image](https://user-images.githubusercontent.com/7318053/147031053-f7af868b-1756-4503-ac8c-a19b14b7a108.png)

7.  Exit Cmder
8. Copy "%userprofile%\cmder_config\user-ConEmu.xml" to "%userprofile%\cmder\vendor\ConEmu.xml.default"
8.  Compress "%userprofile%\cmder" into a zip archive.


Deploying the multi-user redistributable package:

1.  Expand the zip archive to `C:\Program Files\Cmder`
2.  Copy the pre-configured shortcut from `C:\Program Files\Cmder\Cmder.lnk` on the `Start Menu`.

**Comment by gusbemacbe** [comment]
Sorry for my long absence. I have been overcharged the last week, but today I have just sent a complete and well detailed e-mail, explaining everything you said, explained and gave "how-to" to the support analyst and the EPM _squad_. 

I'll return their answer for you.

**Comment by daxgames** [comment]
@gusbemacbe did you find a solution to your questions?

**Comment by gusbemacbe** [comment]
> @gusbemacbe did you find a solution to your questions?

- The security and EPM team authorised to include Cmder file to the Confluence site and said it does not require administrator privileges, but it will not be shared globally among all developers;
- The Cmder has been deployed to the bank software central last week, but the central failed to install it. I have to ask the support analyst to show me code and the log.

**Comment by gusbemacbe** [comment]
Hello!

Good news. 

Cmder was deployed without to the bank software central. I have tested another time, it was installed with success. I open and analysed, seeing the user aliases are generated automatically by a bank script. Only Git for Windows can not be installed due to the bank proxy. But Git Bash is already at the bank software central. We'll write the documentation for bank colleagues to readjust and reintegrate the shells, including Git Bash.

**Comment by daxgames** [comment]
If you install git bash using your normal process Cmder should automatically find it and use it if git is added to the path.
